streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -184,7 +184,11 @@ public void initialize() {
 
             if (exception == null) {
                 final TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
-                offsets.put(tp, metadata.offset());
+                if (metadata.offset() >= 0L) {
+                    offsets.put(tp, metadata.offset());
+                } else {
+                    log.warn("Received offset={} in produce response for {}", metadata.offset(), tp);
```

Comment:
Skip writing/overwriting the offset for this partition, at worst we will have to restore the last batch of records unnecessarily.

Reply:
Did you observe that offset returned from the param could be smaller than zero? I think this should never happen from producer callbacks.

Reply:
No, but Jason said he thought it may be possible in some edge (but non-error) cases. There's no guarantee a successful send will never return a negative offset

Reply:
Cool